Vue entrypoint

farmaazon opened the pull request:
### Pull Request Description

Fixes #11185 
Fixes #11923

I moved (and simplified) some setup. Mostly, the entry point sets up general behavior on page load, App.vue is the root, and the react-specific setups are now in `ReactRoot` component. The old App.vue is now just the ProjectView.vue.

Removed some options which are no longer relevant since there's just a single GUI package.

I tried to put the ProjectView component into dashboard layout, but unfortunately, veaury is not helpful here, as it unsets all styles in its wrapper.

Also, GUI no longer displays Help view; if there's an unrecognized option we just print a warning and continue.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- ~~[ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- ~~[ ] Unit tests have been written where possible.~~
- ~~[ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.~~
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

farmaazon committed:
Change entry point. Tests still not working.

farmaazon committed:
Fix tests

farmaazon committed:
Fixes after rebase

farmaazon committed:
fix prettier

farmaazon committed:
Fix linter WIP

farmaazon committed:
Fix tests

farmaazon committed:
Refactoring

farmaazon committed:
WIP

farmaazon committed:
Fix lint error

farmaazon commented on app/gui/src/entrypoint.ts:
```diff
@@ -1,57 +1,154 @@
+  })
+}
+
+function imNotSureButPerhapsFixingRefreshingWithAuthentication() {
```
@somebody1234 I don't know what exactly the snippet below fixes. The Note is not clear to me.
I tracked that you are the original author, could you explain (and make a better name for this method :) )

farmaazon committed:
Self-review

MrFlashAccount commented on app/gui/src/ReactRoot.tsx:
```diff
@@ -0,0 +1,84 @@
+    //  `style="all: unset"` breaking our layout.
+    //  See https://github.com/gloriasoft/veaury/issues/158
+    <div
+      id="enso-dashboard"
```
Do we need `enso-dashboard` wrapper though? We use this selector only in a few places - define base styles, and in utils, where we detect whenever an element should be hidden by click: `app/gui/src/dashboard/components/AriaComponents/Dialog/utilities.ts`

Guess we can safely migrate from `enso-dashboard` to `enso-app` and remove the wrapper

MrFlashAccount commented on app/gui/src/ReactRoot.tsx:
```diff
@@ -0,0 +1,84 @@
+    //  See https://github.com/gloriasoft/veaury/issues/158
+    <div
+      id="enso-dashboard"
+      className={`${['App', 'enso-dashboard', ...props.classSet.keys()].join(' ')}`}
```
What's these classes stand for?

farmaazon replied:
We have a context allowing just add any class to application root, so far used only when dragging elements.

MrFlashAccount commented on app/gui/src/dashboard/layouts/Editor.tsx:
```diff
@@ -244,16 +196,18 @@ function EditorInternal(props: EditorInternalProps) {
   // Currently the GUI component needs to be fully rerendered whenever the project is changed. Once
   // this is no longer necessary, the `key` could be removed.
-  return AppRunner == null ? null : <AppRunner key={appProps.projectId} {...appProps} />
+  return <ProjectView key={key} {...appProps} />
```
For the future: can we wrap the editor into [KeepAlive](https://vuejs.org/guide/built-ins/keep-alive.html) ?

We're looking into unmounting dom nodes when user switch tabs

farmaazon replied:
I'm not sure if KeepAlive would do what you want? Its docs say about persisting _dynamic_ component (i.e. an instance inside vue's <component>) - so if the tabs would not be rewritten to vue, the component would still be lost.

MrFlashAccount replied:
We'll pass the `hidden` prop to vue and won't unmout the tab content (this is what we do now). So unmounting will be on the vue side, within the `<KeepAlive/>` boundary

MrFlashAccount reviewed (approved):
LGTM, left a few comments

farmaazon committed:
Apply review

farmaazon commented:
Addressed the comments + applied a hack discovered with @Frizi avoiding issues with veaury wrappers (and allowing reducing number of position: absolutes).

farmaazon committed:
Fix flaky test

Frizi commented on app/gui/src/App.vue:
```diff
@@ -0,0 +1,80 @@
+<template>
+  <Teleport to="body">
+    <TooltipDisplayer :registry="appTooltips" />
+  </Teleport>
```
Do we still need this teleport? The tooltip displayer could probably be rendered just at the bottom of this template.

Frizi commented on app/gui/src/dashboard/layouts/Editor.tsx:
```diff
@@ -1,67 +1,33 @@
-
-const IGNORE_PARAMS_REGEX = new RegExp(`^${appUtils.SEARCH_PARAMS_PREFIX}(.+)$`)
+// eslint-disable-next-line no-restricted-syntax
+const ProjectViewTab = applyPureVueInReact(ProjectViewTabVue) as (
+  props: ProjectViewTabProps,
+) => JSX.Element
```
Why is this cast necessary?

farmaazon replied:
Otherwise I get a typecheck error.
```
src/dashboard/layouts/Editor.tsx:212:11 - error TS2786: 'ProjectViewTab' cannot be used as a JSX component.
  Its type 'Function' is not a valid JSX element type.

212   return <ProjectViewTab key={key} {...appProps} />
              ~~~~~~~~~~~~~~
```
JSX element cannot be just any `Function`, and this is what is returned from veaury. If you know better way to fix this error, please let me know.

Frizi replied:
That's annoying, but I guess fine for now. It looks like a typing bug in veaury.

Frizi commented on app/gui/src/project-view/ProjectView.vue:
```diff
@@ -0,0 +1,139 @@
+  cursor: pointer;
+}
+
+:deep([data-use-vue-component-wrap]) {
+  display: contents !important;
+}
```
This is a repeat of style from App.

Frizi commented on app/gui/src/project-view/ProjectViewTab.vue:
```diff
@@ -0,0 +1,23 @@
+
+<template>
+  <KeepAlive>
+    <component :is="hidden ? null : ProjectView" v-bind="projectViewProps"></component>
```
Could this be simplified to this?
```
<ProjectView v-if="!hidden" v-bind="projectViewProps" />
```

Frizi commented on app/gui/src/project-view/stores/project/index.ts:
```diff
@@ -112,36 +105,29 @@ export const [provideProjectStore, useProjectStore] = createContextStore(
     const doc = new Y.Doc()
     const awareness = new Awareness(doc)
 
-    const config = injectGuiConfig()
```
YES thank you! This really needed to go.

Frizi commented on app/gui/src/project-view/stores/project/index.ts:
```diff
@@ -103,7 +89,14 @@ export type ProjectStore = ReturnType<typeof useProjectStore>
 export const [provideProjectStore, useProjectStore] = createContextStore(
   'project',
-  (props: { projectId: string; renameProject: (newName: string) => void }) => {
+  (props: {
+    projectId: string
+    projectName: string
+    projectDisplayedName: string
+    projectNamespace?: string | undefined
+    renameProject: (newName: string) => void
+    engine: LsUrls
+  }) => {
```
I think it's time to move that type to a separate `interface`.

farmaazon committed:
Apply review

farmaazon committed:
Apply review 2

farmaazon committed:
Add a comment
